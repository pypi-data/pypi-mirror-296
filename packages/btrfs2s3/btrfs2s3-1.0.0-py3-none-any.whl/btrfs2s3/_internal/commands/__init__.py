"""Code related to CLI commands for btrfs2s3."""
