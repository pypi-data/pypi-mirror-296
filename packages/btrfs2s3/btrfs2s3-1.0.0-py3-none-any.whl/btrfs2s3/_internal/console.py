"""The global rich text console."""

from __future__ import annotations

from rich.console import Console
from rich.style import Style
from rich.theme import Theme

STYLE_KEY = Style.parse("green")
STYLE_KEEP = Style.parse("green")
STYLE_NOT_KEEPING = Style.parse("bold bright_red")
STYLE_CTRANSID = Style.parse("cyan")

STYLE_CREATE = Style.parse("green")
STYLE_MODIFY = Style.parse("green")
STYLE_DELETE = Style.parse("red")

THEME = Theme(
    {
        "key": STYLE_KEY,
        "keep": STYLE_KEEP,
        "not_keeping": STYLE_NOT_KEEPING,
        "ctransid": STYLE_CTRANSID,
        "create": STYLE_CREATE,
        "modify": STYLE_MODIFY,
        "delete": STYLE_DELETE,
    }
)

CONSOLE = Console(theme=THEME)
"""The global rich text console."""
