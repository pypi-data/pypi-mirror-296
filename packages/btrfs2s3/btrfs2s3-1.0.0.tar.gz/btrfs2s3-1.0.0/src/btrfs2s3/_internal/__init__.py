"""The implementation for the btrfs2s3 command line interface."""
