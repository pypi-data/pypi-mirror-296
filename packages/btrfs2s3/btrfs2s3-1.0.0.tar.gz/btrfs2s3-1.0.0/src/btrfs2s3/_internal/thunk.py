"""A Thunk wraps a value that might be lazily evaluated."""

from __future__ import annotations

import enum
from functools import total_ordering
from typing import Callable
from typing import Generic
from typing import Literal
from typing import overload
from typing import Protocol
from typing import runtime_checkable
from typing import Union

from typing_extensions import TypeAlias
from typing_extensions import TypeVar


class Error(Exception):
    """The top-level class for errors produced by this module."""


class StillTBDError(Error):
    """Thunk.check() was called, but the thunk hasn't been evaluated yet."""


_T = TypeVar("_T")
_T_contra = TypeVar("_T_contra", contravariant=True)


@runtime_checkable
class _SupportsLT(Protocol[_T_contra]):
    def __lt__(self, other: _T_contra) -> bool: ...


class _TbdType(enum.Enum):
    TBD = enum.auto()

    def __str__(self) -> str:
        return "TBD"


TBD = _TbdType.TBD
"""A sentinel value, meaning the real value hasn't been determined yet."""


@total_ordering
class Thunk(Generic[_T]):
    """A Thunk is a wrapper for a value which may or may not be lazily computed.

    A pre-computed Thunk will return its pre-computed value from peek(),
    check() and __call__().

    A lazily-evaluated Thunk will return the TBD sentinel value from peek(),
    and will raise StillTBDError from check(). __call__() will run the lazy
    evaluation function, and save the value. After this the Thunk will behave
    exactly like a pre-computed Thunk. The evaluation function will not be run
    again.

    This is a very rudimentary lazy evaluation wrapper class. Its only goal is
    to facilitate creating a "dry-run plan" for code that does destructive
    changes, and to present that to the user.

    The intended use of Thunk is as a building block to create complex "intent"
    objects, e.g. an intent to create a backup of some data. The various
    parameters (data source, storage location, etc) of this intent can be
    stored as Thunks which may be precomputed
    or lazily evaluated. This intent can be previewed to a user, and optionally
    executed. Using Thunks allows the preview and execution code to be agnostic
    of data dependencies and edge cases. The preview code can just display all
    the precomputed values in the intent object, and the execution code can
    just evaluate the thunks as needed. Note that this means all the complexity
    is pushed to the code that creates the intent objects.
    """

    _value: _T | Literal[_TbdType.TBD]
    _get_value: Callable[[], _T]

    # Without these overloads, mypy can't infer the type of Thunk(other_thunk),
    # presumably it can't figure out if this should match __init__(_T) or
    # __init__(Thunk[_T]).
    @overload
    def __init__(self, other: Thunk[_T]) -> None: ...
    @overload
    def __init__(self, other: Callable[[], _T]) -> None: ...
    @overload
    def __init__(self, other: _T) -> None: ...
    def __init__(self, other: Thunk[_T] | Callable[[], _T] | _T) -> None:
        """Create a thunk.

        Thunk(other_thunk) creates a copy of other_thunk. The result will have
        the same precomputed value and/or evaluation function as the argument
        Thunk.

        Thunk(callable) creates a lazily-evaluated Thunk. The argument must be
        a zero-args callable.

        Thunk(any-other-value) creates a pre-computed Thunk.

        Args:
            other: A zero-args callable, or another Thunk, or any other
                precomputed value.
        """
        if isinstance(other, Thunk):
            self._get_value = other._get_value  # noqa: SLF001
            self._value = other._value  # noqa: SLF001
        elif callable(other):
            self._get_value = other
            self._value = TBD
        else:
            self._get_value = lambda: other  # pragma: no cover
            self._value = other

    def peek(self) -> _T | Literal[_TbdType.TBD]:
        """Returns the precomputed value, or the TBD sentinel value."""
        return self._value

    def check(self) -> _T:
        """Returns the precomputed value, or raises StillTBDError.

        This asserts that the Thunk is pre-computed. It is mostly useful in
        tests.

        Raises:
            StillTBDError: If the thunk is lazy, and hasn't been evaluated yet.
        """
        value = self._value
        if value is TBD:
            raise StillTBDError
        return value

    def is_tbd(self) -> bool:
        """Returns True if the Thunk is lazy and hasn't been evaluated yet."""
        return self._value is TBD

    def __call__(self) -> _T:
        """Returns the value of the Thunk, possibly running its lazy computation.

        Returns:
            The value of the Thunk.
        """
        if self._value is TBD:
            self._value = self._get_value()
        return self._value

    def __repr__(self) -> str:
        """Return a string representation of the Thunk."""
        value = self._value
        if value is TBD:
            return "Thunk(TBD)"
        return f"Thunk({value!r})"

    def __lt__(self, other: Thunk[_T]) -> bool:
        """Returns whether this Thunk should appear before other Thunks.

        The natural ordering of Thunks preserves the natural ordering of their
        values, except that lazy Thunks appear after precomputed Thunks.
        """
        value = self._value
        othervalue = other._value
        if value is not TBD:
            if othervalue is not TBD:
                if not isinstance(value, _SupportsLT):
                    return NotImplemented
                return value.__lt__(othervalue)
            return True
        if othervalue is not TBD:
            return False
        return id(self) < id(other)


ThunkArg: TypeAlias = Union[Thunk[_T], _T, Callable[[], _T]]
"""An alias to allowed types for the argument to Thunk()."""
