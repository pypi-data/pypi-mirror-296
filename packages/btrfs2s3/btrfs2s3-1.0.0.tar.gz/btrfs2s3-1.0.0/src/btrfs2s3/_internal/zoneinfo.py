"""Compatibility module for zoneinfo."""

import sys

if sys.version_info >= (3, 9):  # pragma: >=3.9 cover
    from zoneinfo import ZoneInfo
else:  # pragma: <3.9 cover
    from backports.zoneinfo import ZoneInfo


# This function is just syntactic sugar to call the ZoneInfo constructor with
# appropriate lint markers
def get_zoneinfo(key: str) -> ZoneInfo:
    """Return a ZoneInfo object from a string timezone name."""
    if sys.version_info >= (3, 9):  # pragma: >=3.9 cover
        return ZoneInfo(key)
    else:  # noqa: RET505 # pragma: <3.9 cover
        # https://github.com/pganssle/zoneinfo/issues/125
        # Unfortunately the workaround in the linked issue does not work with
        # mypy --strict (I get: Call to untyped function "ZoneInfo" in typed
        # context)

        # Also mypy requires if-return-else to understand the version
        # condition, which thus requires the RET505 block. Yakkity yak
        return ZoneInfo(key)  # type: ignore[abstract]
