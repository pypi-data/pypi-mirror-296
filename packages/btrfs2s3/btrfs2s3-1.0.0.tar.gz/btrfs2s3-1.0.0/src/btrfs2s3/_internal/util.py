from __future__ import annotations

from enum import IntFlag

from btrfsutil import SubvolumeInfo

from btrfs2s3._internal.backups import BackupInfo

NULL_UUID = b"\0" * 16


def mksubvol(
    *,
    id: int = 0,  # noqa: A002
    parent_id: int = 0,
    dir_id: int = 0,
    flags: int = 0,
    uuid: bytes = NULL_UUID,
    parent_uuid: bytes = NULL_UUID,
    received_uuid: bytes = NULL_UUID,
    generation: int = 0,
    ctransid: int = 0,
    otransid: int = 0,
    stransid: int = 0,
    rtransid: int = 0,
    ctime: float = 0.0,
    otime: float = 0.0,
    stime: float = 0.0,
    rtime: float = 0.0,
) -> SubvolumeInfo:
    return SubvolumeInfo(
        (
            id,
            parent_id,
            dir_id,
            flags,
            uuid,
            parent_uuid,
            received_uuid,
            generation,
            ctransid,
            otransid,
            stransid,
            rtransid,
            ctime,
            otime,
        ),
        {"stime": stime, "rtime": rtime},
    )


def backup_of_snapshot(
    snapshot: SubvolumeInfo, send_parent: SubvolumeInfo | None = None
) -> BackupInfo:
    return BackupInfo(
        uuid=snapshot.uuid,
        parent_uuid=snapshot.parent_uuid,
        send_parent_uuid=None if send_parent is None else send_parent.uuid,
        ctransid=snapshot.ctransid,
        ctime=snapshot.ctime,
    )


class SubvolumeFlags(IntFlag):
    ReadOnly = 1
    Proposed = 2**30
