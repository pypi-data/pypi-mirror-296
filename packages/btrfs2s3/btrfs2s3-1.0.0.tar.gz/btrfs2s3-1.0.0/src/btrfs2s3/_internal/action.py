"""Actions that modify snapshots or backups."""

from __future__ import annotations

import dataclasses
import errno
from itertools import chain
import logging
from subprocess import PIPE
from subprocess import Popen
from typing import TYPE_CHECKING

import btrfsutil

from btrfs2s3._internal.stream_uploader import upload_non_seekable_stream_via_tempfile
from btrfs2s3._internal.thunk import Thunk
from btrfs2s3._internal.thunk import ThunkArg
from btrfs2s3._internal.util import NULL_UUID
from btrfs2s3._internal.util import SubvolumeFlags

if TYPE_CHECKING:
    from pathlib import Path
    from typing import Iterator
    from typing import Sequence

    from mypy_boto3_s3.client import S3Client

_LOG = logging.getLogger(__name__)

DEFAULT_PART_SIZE = 5 * 2**30


def create_snapshot(*, source: Path, path: Path) -> None:
    """Create a read-only snapshot of a subvolume.

    Args:
        source: The source subvolume.
        path: The path at which to create a read-only snapshot.
    """
    _LOG.info("creating read-only snapshot of %s at %s", source, path)
    btrfsutil.create_snapshot(source, path, read_only=True)


def delete_snapshot(path: Path) -> None:
    """Delete a read-only snapshot of a subvolume.

    Args:
        path: The path to a read-only snapshot to be deleted.

    Raises:
        RuntimeError: If one of the arguments does not refer to a read-only
            snapshot of a subvolume.
    """
    # Do some extra checks to make sure we only ever delete read-only
    # snapshots, not source subvolumes.
    if not btrfsutil.is_subvolume(path):
        msg = "target isn't a subvolume"
        raise RuntimeError(msg)
    info = btrfsutil.subvolume_info(path)
    if info.parent_uuid == NULL_UUID:
        msg = "target isn't a snapshot"
        raise RuntimeError(msg)
    if not info.flags & SubvolumeFlags.ReadOnly:
        msg = "target isn't a read-only snapshot"
        raise RuntimeError(msg)
    _LOG.info("deleting read-only snapshot %s", path)
    try:
        btrfsutil.delete_subvolume(path)
    except OSError as ex:
        if ex.errno == errno.EROFS:
            # https://github.com/sbrudenell/btrfs2s3/issues/38
            # If we set the subvolume read-write but fail to delete it, later
            # runs will ignore it, creating a leak. To minimize this, only do
            # it in the non-root case.
            btrfsutil.set_subvolume_read_only(path, read_only=False)
            btrfsutil.delete_subvolume(path)
        else:
            raise  # pragma: no cover


def rename_snapshot(*, source: Path, target: Path) -> None:
    """Rename a read-only snapshot of of a subvolume.

    Args:
        source: The source snapshot.
        target: The new path of the snapshot.
    """
    _LOG.info("renaming %s -> %s", source, target)
    source.rename(target)


def create_backup(
    *,
    s3: S3Client,
    bucket: str,
    snapshot: Path,
    send_parent: Path | None,
    key: str,
    pipe_through: Sequence[Sequence[str]] = (),
    part_size: int = DEFAULT_PART_SIZE,
) -> None:
    """Stores a btrfs archive in S3.

    This will spawn "btrfs -q send" as a subprocess, as there is currently no way
    to create a btrfs-send stream via pure python.

    This will temporarily store the "btrfs send" stream on disk. A maximum of
    part_size bytes will be stored at a time.

    Args:
        s3: An S3 client.
        bucket: The bucket in which to store the archive.
        snapshot: The snapshot to back up.
        send_parent: The parent snapshot for delta backups. When not None, this
            will be supplied to "btrfs send -p".
        key: The S3 object key.
        pipe_through: A sequence of shell commands through which the archive
            should be piped before uploading.
        part_size: For multipart uploads, use this as the maximum part size.
            Defaults to the well-known maximum part size for AWS, which is
            currently 5 GiB.
    """
    _LOG.info(
        "creating backup of %s (%s)",
        snapshot,
        f"delta from {send_parent}" if send_parent else "full",
    )
    send_args: list[str | Path] = ["btrfs", "send", "-q"]
    if send_parent is not None:
        send_args += ["-p", send_parent]
    send_args += [snapshot]

    pipeline: list[Popen[bytes]] = []
    for args in chain((send_args,), pipe_through):
        prev_stdout = pipeline[-1].stdout if pipeline else None
        pipeline.append(Popen(args, stdin=prev_stdout, stdout=PIPE))  # noqa: S603
        # https://docs.python.org/3/library/subprocess.html#replacing-shell-pipeline
        if prev_stdout:
            prev_stdout.close()

    pipeline_stdout = pipeline[-1].stdout
    # https://github.com/python/typeshed/issues/3831
    assert pipeline_stdout is not None  # noqa: S101
    try:
        upload_non_seekable_stream_via_tempfile(
            client=s3,
            bucket=bucket,
            key=key,
            stream=pipeline_stdout,
            part_size=part_size,
        )
    finally:
        # Allow the pipeline to fail if the upload fails
        pipeline_stdout.close()

    # reverse order to match the semantics of pipefail
    for process in reversed(pipeline):
        if process.wait() != 0:
            msg = f"{process.args!r}: exited with code {process.returncode}"
            try:
                raise RuntimeError(msg)
            finally:
                # Assume the backup is corrupted
                delete_backups(s3, bucket, key)


def delete_backups(s3: S3Client, bucket: str, *keys: str) -> None:
    """Batch delete backups from S3.

    This will use the DeleteObjects API call, which can delete multiple keys in
    batches.

    Args:
        s3: An S3 client.
        bucket: The bucket from which to delete keys.
        *keys: The keys to delete.
    """

    def batches() -> Iterator[tuple[str, ...]]:
        for i in range(0, len(keys), 1000):
            yield keys[i : i + 1000]

    for batch in batches():
        for key in batch:
            _LOG.info("deleting backup %s", key)
        # Do we need to inspect the response for individual errors, or will we
        # raise an exception in this case? The docs are thousands of words long
        # but don't explain this
        s3.delete_objects(
            Bucket=bucket,
            Delete={"Quiet": True, "Objects": [{"Key": key} for key in batch]},
        )


# Future: all the fields of the intent objects should really be
# descriptor-typed fields that convert things to Thunk.


@dataclasses.dataclass(frozen=True, order=True)
class CreateSnapshot:
    """An intent to create a read-only snapshot of a subvolume."""

    source: Thunk[Path]
    path: Thunk[Path]


@dataclasses.dataclass(frozen=True, order=True)
class DeleteSnapshot:
    """An intent to delete a read-only snapshot."""

    path: Thunk[Path]


@dataclasses.dataclass(frozen=True, order=True)
class RenameSnapshot:
    """An intent to rename a read-only snapshot."""

    source: Thunk[Path]
    target: Thunk[Path]


@dataclasses.dataclass(frozen=True, order=True)
class CreateBackup:
    """An intent to create a backup of a read-only snapshot."""

    source: Thunk[Path]
    snapshot: Thunk[Path]
    send_parent: Thunk[Path | None]
    key: Thunk[str]


@dataclasses.dataclass(frozen=True, order=True)
class DeleteBackup:
    """An intent to delete a backup."""

    key: Thunk[str]


class Actions:
    """A list of actions to be executed.

    This class just stores a list of intents. Intents of actions can be added
    with the various functions, and then examined with the iter_*_intents()
    functions. It doesn't actually perform any actions until execute() is called.

    All of the arguments to the intent functions are converted to Thunks.

    Note that when creating Thunks, it's critical to understand the order that
    actions are performed. See execute().
    """

    def __init__(self) -> None:
        """Construct an empty Actions object."""
        self._create_snapshots: list[CreateSnapshot] = []
        self._delete_snapshots: list[DeleteSnapshot] = []
        self._rename_snapshots: list[RenameSnapshot] = []
        self._create_backups: list[CreateBackup] = []
        self._delete_backups: list[DeleteBackup] = []

    def empty(self) -> bool:
        """Returns True if there are no intended actions."""
        return not (
            self._create_snapshots
            or self._delete_snapshots
            or self._rename_snapshots
            or self._create_backups
            or self._delete_backups
        )

    def create_snapshot(self, *, source: ThunkArg[Path], path: ThunkArg[Path]) -> None:
        """Add an intent to create a read-only snapshot of a subvolume.

        Args:
            source: The path to the source subvolume, of which a snapshot
                should be created.
            path: The path where the new snapshot should be created.
        """
        self._create_snapshots.append(
            CreateSnapshot(source=Thunk(source), path=Thunk(path))
        )

    def rename_snapshot(
        self, *, source: ThunkArg[Path], target: ThunkArg[Path]
    ) -> None:
        """Add an intent to rename a snapshot.

        Args:
            source: The initial path of the snapshot.
            target: The new path of the snapshot.
        """
        self._rename_snapshots.append(
            RenameSnapshot(source=Thunk(source), target=Thunk(target))
        )

    def delete_snapshot(self, path: ThunkArg[Path]) -> None:
        """Add an intent to delete a snapshot.

        Args:
            path: The path of the snapshot to be deleted.
        """
        self._delete_snapshots.append(DeleteSnapshot(path=Thunk(path)))

    def create_backup(
        self,
        *,
        source: ThunkArg[Path],
        snapshot: ThunkArg[Path],
        send_parent: ThunkArg[Path | None],
        key: ThunkArg[str],
    ) -> None:
        """Add an intent to create a new backup of a read-only snapshot.

        The backup will be created using `btrfs send` (if send_parent is not
        None, then `btrfs send -p` will be used). The backup will be streamed
        into an S3 object with the given key.

        Args:
            source: The path to the source subvolume of the snapshot. This
                has no effect when creating the backup, but it's very useful
                when examining intents with iter_create_backup_intents().
            snapshot: The path to the read-only snapshot to be backed up.
            send_parent: The path to a read-only snapshot to be used as a
                parent (will be passed to `btrfs send -p`).
            key: The S3 object key at which to store the backup.
        """
        self._create_backups.append(
            CreateBackup(
                source=Thunk(source),
                snapshot=Thunk(snapshot),
                send_parent=Thunk(send_parent),
                key=Thunk(key),
            )
        )

    def delete_backup(self, key: ThunkArg[str]) -> None:
        """Add an intent to delete an S3 object.

        The target S3 object should ostensibly be a backup previously created,
        but we don't verify this.

        Args:
            key: The S3 object key to be deleted.
        """
        self._delete_backups.append(DeleteBackup(Thunk(key)))

    def iter_create_snapshot_intents(self) -> Iterator[CreateSnapshot]:
        """Iterates all the CreateSnapshot intents."""
        yield from sorted(self._create_snapshots)

    def iter_delete_snapshot_intents(self) -> Iterator[DeleteSnapshot]:
        """Iterates all the DeleteSnapshot intents."""
        yield from sorted(self._delete_snapshots)

    def iter_rename_snapshot_intents(self) -> Iterator[RenameSnapshot]:
        """Iterates all the RenameSnapshot intents."""
        yield from sorted(self._rename_snapshots)

    def iter_create_backup_intents(self) -> Iterator[CreateBackup]:
        """Iterates all the CreateBackup intents."""
        yield from sorted(self._create_backups)

    def iter_delete_backup_intents(self) -> Iterator[DeleteBackup]:
        """Iterates all the DeleteBackup intents."""
        yield from sorted(self._delete_backups)

    def execute(
        self, s3: S3Client, bucket: str, pipe_through: Sequence[Sequence[str]] = ()
    ) -> None:
        """Executes the intended actions.

        This performs all the side effects described in the various intent
        objects. It will create/rename/delete snapshots using btrfsutil, and
        create/delete backups in the supplied S3 bucket.

        The actions are performed in the following order:

        - Create snapshots
        - Rename snapshots
        - Create backups
        - Delete snapshots
        - Delete backups

        Within an action type, the actions are performed in the same order
        returned from the iter_*_intents() functions. That is, snapshots are
        created in the same order that they are returned from
        iter_create_snapshot_intents(), etc.

        Args:
            s3: The S3 client object to use to manipulate S3 objects.
            bucket: The name of the bucket where backups are stored.
            pipe_through: A sequence of shell commands through which backup
                archives should be piped before uploading.
        """
        for create_snapshot_intent in self.iter_create_snapshot_intents():
            create_snapshot(
                source=create_snapshot_intent.source(),
                path=create_snapshot_intent.path(),
            )

        for rename_snapshot_intent in self.iter_rename_snapshot_intents():
            rename_snapshot(
                source=rename_snapshot_intent.source(),
                target=rename_snapshot_intent.target(),
            )

        for create_backup_intent in self.iter_create_backup_intents():
            create_backup(
                s3=s3,
                bucket=bucket,
                snapshot=create_backup_intent.snapshot(),
                send_parent=create_backup_intent.send_parent(),
                key=create_backup_intent.key(),
                pipe_through=pipe_through,
            )

        for delete_snapshot_intent in self.iter_delete_snapshot_intents():
            delete_snapshot(delete_snapshot_intent.path())

        keys = tuple(d.key() for d in self.iter_delete_backup_intents())
        delete_backups(s3, bucket, *keys)
