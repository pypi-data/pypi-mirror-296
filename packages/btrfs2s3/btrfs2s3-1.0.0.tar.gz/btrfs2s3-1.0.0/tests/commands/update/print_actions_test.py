from __future__ import annotations

from pathlib import Path
from typing import TYPE_CHECKING

from btrfs2s3._internal.action import Actions
from btrfs2s3._internal.commands.update import print_actions
import pytest

if TYPE_CHECKING:
    from rich.console import Console


@pytest.mark.parametrize(
    (
        "n_create_snapshots",
        "n_rename_snapshots",
        "n_create_backups",
        "n_delete_snapshots",
        "n_delete_backups",
    ),
    [(0, 0, 0, 0, 0), (3, 3, 3, 3, 3)],
)
def test_print_actions(
    goldifyconsole: Console,
    n_create_snapshots: int,
    n_rename_snapshots: int,
    n_create_backups: int,
    n_delete_snapshots: int,
    n_delete_backups: int,
) -> None:
    actions = Actions()
    for i in range(n_create_snapshots):
        actions.create_snapshot(
            source=Path(f"/path/to/source{i}"), path=Path(f"/path/to/snapshot{i}")
        )
    for i in range(n_rename_snapshots):
        actions.rename_snapshot(
            source=Path(f"/path/to/source{i}"), target=Path(f"/path/to/target{i}")
        )
    for i in range(n_create_backups):
        actions.create_backup(
            source=Path(f"/path/to/source{i}"),
            snapshot=Path(f"/path/to/snapshot{i}"),
            key=f"key{i}",
            send_parent=Path(f"/path/to/send/parent{i}"),
        )
    for i in range(n_delete_snapshots):
        actions.delete_snapshot(Path(f"/path/to/snapshot{i}"))
    for i in range(n_delete_backups):
        actions.delete_backup(f"key{i}")

    print_actions(console=goldifyconsole, actions=actions)
