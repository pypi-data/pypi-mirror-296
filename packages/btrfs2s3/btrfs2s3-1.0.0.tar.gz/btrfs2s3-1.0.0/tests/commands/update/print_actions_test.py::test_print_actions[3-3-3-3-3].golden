[92m───────────────────────────────── [0mactions to be taken[92m ──────────────────────────────────[0m
[32m [0m[32m📸 create snapshot:[0m[32m [0m[32m [0m[32m/path/to/source0[0m[32m [0m[32m [0m[32m->[0m[32m [0m[32m [0m[32m/path/to/snapshot0[0m[32m [0m
[2;32m [0m[2;32m📸 create snapshot:[0m[2;32m [0m[2;32m [0m[2;32m/path/to/source1[0m[2;32m [0m[2;32m [0m[2;32m->[0m[2;32m [0m[2;32m [0m[2;32m/path/to/snapshot1[0m[2;32m [0m
[32m [0m[32m📸 create snapshot:[0m[32m [0m[32m [0m[32m/path/to/source2[0m[32m [0m[32m [0m[32m->[0m[32m [0m[32m [0m[32m/path/to/snapshot2[0m[32m [0m

[32m [0m[32m📝️ rename snapshot:[0m[32m [0m[32m [0m[32m/path/to/source0[0m[32m [0m[32m [0m[32m->[0m[32m [0m[32m [0m[32m/path/to/target0[0m[32m [0m
[2;32m [0m[2;32m📝️ rename snapshot:[0m[2;32m [0m[2;32m [0m[2;32m/path/to/source1[0m[2;32m [0m[2;32m [0m[2;32m->[0m[2;32m [0m[2;32m [0m[2;32m/path/to/target1[0m[2;32m [0m
[32m [0m[32m📝️ rename snapshot:[0m[32m [0m[32m [0m[32m/path/to/source2[0m[32m [0m[32m [0m[32m->[0m[32m [0m[32m [0m[32m/path/to/target2[0m[32m [0m

[32m [0m[32m☁️ upload backup:[0m[32m [0m[32m [0m[32m/path/to/snapshot0[0m[32m [0m
[2;32m [0m[2;32m☁️ upload backup:[0m[2;32m [0m[2;32m [0m[2;32m/path/to/snapshot1[0m[2;32m [0m
[32m [0m[32m☁️ upload backup:[0m[32m [0m[32m [0m[32m/path/to/snapshot2[0m[32m [0m

[31m [0m[31m💀 delete snapshot:[0m[31m [0m[31m [0m[31m/path/to/snapshot0[0m[31m [0m
[2;31m [0m[2;31m💀 delete snapshot:[0m[2;31m [0m[2;31m [0m[2;31m/path/to/snapshot1[0m[2;31m [0m
[31m [0m[31m💀 delete snapshot:[0m[31m [0m[31m [0m[31m/path/to/snapshot2[0m[31m [0m

[31m [0m[31m💀 delete backup: [0m[31m [0m[31m [0m[31mkey0[0m[31m [0m
[2;31m [0m[2;31m💀 delete backup: [0m[2;31m [0m[2;31m [0m[2;31mkey1[0m[2;31m [0m
[31m [0m[31m💀 delete backup: [0m[31m [0m[31m [0m[31mkey2[0m[31m [0m

